### PR TITLE
haskell.compiler.ghc865Binary: add powerpc64le bootstrap 

### PR DESCRIPTION
**Pull request by ghost**
~~Marked as draft until staging reopens on 2022-May-15.~~

###### Description of changes

This PR adds the ghc 8.6.5 bootstrap binaries for powerpc64le, as recommended by @sternenseemann here:

  https://github.com/NixOS/nixpkgs/pull/168113#issuecomment-1094454646

###### Things done

- Built on platform(s)
  - [X] powerpc64le-linux
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/ghc/8.6.5-binary.nix
Comment (sternenseemann) [inline pkgs/development/compilers/ghc/8.6.5-binary.nix:16]: Otherwise there's an ambiguity w.r.t. endianness, is there not?

```suggestion
                || stdenv.hostPlatform.system == "powerpc64le-linx";
```

**Review by sternenseemann — COMMENTED** [review]
I'll convert this to a draft until #168199 is merged after which this will become useful. Are there plans for CI / testing / Hydra jobset already?

Unfortunately it seems that GHC HQ's powerpc builders broke or were decomissioned some time after 8.6.5, so we can't bootstrap anything after 8.10.7 (GHC only supports the previous two major releases for bootstrapping). We can establish a chain for now, i.e. build ``ghc902`,` `ghc922` and `ghcHEAD` using `ghc8107` (not binary!) on `powerpc64le-linux` for now — that'll work, but be annoying as it requires building two full GHCs.

**Comment by a-m-joseph** [comment]
> I'll convert this to a draft until #168199 is merged

Sigh.  I would appreciate it if you would give an approval-review to that PR then, or at least poke the people who are in a position to merge it.

One of the most dissatisfying parts of my experience with the mips64 support was everybody kind of dodging merges/reviews by making PR X a draft until PR Y merged.  This is especially frustrating since *github refuses to implement any form of bug dependencies*, something bugzilla has been able to do for decades...

In the mips64 case, the uber-blocker bug was this one: https://github.com/NixOS/nixpkgs/pull/166879.  If you can prod people into taking that seriously you'll be my personal hero.  A ton of packages vendored an impurity from libtool, and waiting for all of those packages to re-run autoconf is going take several years.  In the mips64 case it's extra frustrating because the bug also blocks Hydra from generating a bootstrap-files (whereas for powerpc64le we already have that).

> after which this will become useful. Are there plans for CI / testing / Hydra jobset already?

Western Semiconductor offered to donate about $1,000 worth of mips64 hardware to nixorg.  @GrahamC mediated discussions with the nixorg board (or whatever it is called), who initially agreed to accept the donation, but then things sort of went to radio silence.

Anyways I am open to possibly doing something similar for power64le, but since the hardware is about six times as expensive (and about 100 times higher-performance) it would be a longer process to get that approved and may require additional conditions (like agreeing to return the hardware if it falls into disuse).

File: pkgs/development/compilers/ghc/8.6.5-binary.nix
Comment (a-m-joseph) [reply pkgs/development/compilers/ghc/8.6.5-binary.nix:16]: Well, that's intentional... I can't imagine that endianness affects whether or not `ncurses6` should be used.  Frankly I'm not sure under what conditions that variable should be set or why (if you know please tell me and I will add a comment) but it seems to have something to do with 64-bit systems, which is why I set it to `true` for 64-bit powerpc.  Also, it does in fact work correctly... I have `nix-tree` now!  Hooray!

In general we should be moving away from using `builtins.system`; it's sort of `nix`'s ad-hoc platform taxonomy that doesn't represent things like abi, libc.

Also I think you have a typo there with `linx`.  That's another reason to use the predicates from `inspect.nix` rather than string comparison.

**Comment by a-m-joseph** [comment]
> Unfortunately it seems that GHC HQ's powerpc builders broke or were decomissioned some time after 8.6.5, so we can't bootstrap anything after 8.10.7 (GHC only supports the previous two major releases for bootstrapping).

With this commit ghc 8.10.7 built without any problems (compiled by 8.6.5).

> We can establish a chain for now, i.e. build ``ghc902`,` `ghc922` and `ghcHEAD` using `ghc8107` (not binary!) on `powerpc64le-linux` for now — that'll work, but be annoying as it requires building two full GHCs.

I'd like to try building these, but `nix-build -A ghc902` gives me:

```
error: attribute 'ghc902' in selection path 'ghc902' not found
```

Can you explain to me how the attribute-paths for specific versions of ghc are supposed to work?  I struggle to make any sense of them.  Also not sure about `haskell.ghc` vs `haskellPackages.*`...

File: pkgs/development/compilers/ghc/8.6.5-binary.nix
Comment (sternenseemann) [reply pkgs/development/compilers/ghc/8.6.5-binary.nix:16]: Yes, the library used does not depend on on the architecture, but that doesn't mean we can freely choose it! It depends on what version of the library the bindist was linked against before it was released, so it needs to be checked *per bindist* and an assumption based on CPU family won't give you a correct result. (I checked and Fedora 29 which this bindist was built using used to ship ncurses 6.)

Comment (a-m-joseph) [reply pkgs/development/compilers/ghc/8.6.5-binary.nix]: Ah, okay.  Updated to account for this.

We should probably not have a default value here then, and should `throw` if the platform is not one of those for which somebody has manually checked this.

**Comment by a-m-joseph** [comment]
> I'm trying these now.

Pushed a commit to allow ghc9xx to evaluate.  Build in progress.

**Comment by a-m-joseph** [comment]
> Pushed a commit to allow ghc9xx to evaluate. Build in progress.

`ghc902` built successfully.

Now building `ghc921`.

**Comment by sternenseemann** [comment]
> Sigh. I would appreciate it if you would give an approval-review to that PR then, or at least poke the people who are in a position to merge it.

I'll try to look into it soon (if I'm lucky I have access to hardware to test it).

I can sympathize with your frustration. I don't think it's a technical problem (what isn't solved already should be possible to), really, which is why I think everyone is so hesitant. We need to decide whether to add support for this (these) new platform(s) and the discuss how to and to what extent maintain the support.

I must say that we have a really bad track record with decision making, the RFC process is slow and has a tendency to delve into trivial things, and apart from that it kind of hinges on individuals (this is a general problem, e.g. adding machines to Hydra can be done by a low one digit amount of people to my knowledge). For this we don't need an RFC, as there's enough precedent for platforms being added without one, but it also serves as diffusion of responsibility.

Also in my mind I am increasingly hyper-aware that adding support for something is always a risk, i.e. you are forced to keep it in that way afterwards and it is often difficult to gauge, how feasible that'll be in the future (especially with key people leaving, lack of CI / resources for testing / …). This is not to say that I'm not in favor, just to explain my (and maybe other's?) hesitation. With the improved bootstrapping by having the necessary tools cross compiled by Hydra first and the fact that it is Linux platforms we are talking about, it should be pretty safe overall.

If I recall correctly, the riscv64 support was also stuck in limbo for quite some time. I suppose mips and powerpc may also be a tougher sell / spark less interest for people (as opposed to the recent additions relating to aarch64/riscv64).

> One of the most dissatisfying parts of my experience with the mips64 support was everybody kind of dodging merges/reviews by making PR X a draft until PR Y merged.

At least I can assure you that I'm very prepared to merge *this* PR, as it is actually my area of responsibility.

> Western Semiconductor offered to donate about $1,000 worth of mips64 hardware to nixorg with no strings attached. @grahamc mediated discussions with the nixorg board (or whatever it is called), who initially agreed to accept the donation. I asked for a shipping address and then things sort of went to radio silence.
> 
> Anyways I am open to possibly doing something similar for power64le, but since the hardware is about 12 times as expensive (and about 100 times higher-performance) it would be a longer process to get that approved and there would be additional conditions -- like agreeing to return the hardware if it isn't being used.

That sounds pretty exciting.

File: pkgs/top-level/haskell-packages.nix
Comment (sternenseemann) [inline pkgs/top-level/haskell-packages.nix:99]: Could you add a comment for these that there's no upstream bindist available for a bootstrap- suitable GHC version?

**Review by sternenseemann — APPROVED** [review]

File: pkgs/development/compilers/ghc/8.6.5-binary.nix
Comment (sternenseemann) [reply pkgs/development/compilers/ghc/8.6.5-binary.nix]: Probably, yeah. The newer bindists >= 8.10.2 have a better logic for this, but 8.6.5 was never touched, as it's simple and we thought it wouldn't stick around for too long.

**Comment by sternenseemann** [comment]
Oh, and the commit message(s) should read `haskell.compiler.ghc865Binary: …` for the bindist commits when you squash at the end.

**Comment by a-m-joseph** [comment]
> Now building `ghc921`.

Completed successfully.

Have to leave the office for a bit, will reply to the three comments above when I get back.

**Comment by sternenseemann** [comment]
> > Now building `ghc921`.
> 
> Completed successfully.

To clarify, you mean `ghc922`, right? `gh921` was removed a few weeks ago on `master`.

File: pkgs/development/compilers/ghc/8.6.5-binary.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/ghc/8.6.5-binary.nix:78]: Version numbers and the domain should probably be substituted like above.

Comment (a-m-joseph) [reply pkgs/development/compilers/ghc/8.6.5-binary.nix]: Fixed and squashed.

**Comment by a-m-joseph** [comment]
> We need to decide whether to add support for this (these) new platform(s) and the discuss how to and to what extent maintain the support.

Using the word "support" is really misleading here.

It's an unfortunate quirk of the English language where the same noun is used to describe both (a) the burden of assisting users on a given platform and (b) merging changes which allow compilation to succeed on a platform.  I wish we had different words for these two things, all I want is the latter, so these changes don't have to be kept out of tree where they will get merge-conflicted by every treewide or typo on an adjacent line.

> Also in my mind I am increasingly hyper-aware that adding support for something is always a risk

I really don't see things this way for adding new platforms, especially for cases like mips64 and powerpc64el where 95%+ of the changes involve adding another line to an existing "switch(architecture) { ... }" block (well, the nixlang equivalent of that).

> If I recall correctly, the riscv64 support was also stuck in limbo for quite some time. I suppose mips and powerpc may also be a tougher sell / spark less interest for people

I bristle a bit at this.

I can order mips64 routers off of Amazon, today -- commercially profitable products with have multiple gigabit ethernet ports and optical fiber SFP cages, and I can put 8GB of ram into them.  Riscv64 is still a mishmash of sold-at-a-loss dev boards and all but one have soldered-down ram inadequate to build nixpkgs.

I'm sure someday riscv will supplant MIPS.  It would've been called "MIPS VI" if not for trademarks -- Dave Patterson and his collaborators created both of them!  But people are ridiculously optimistic about how soon that's going to happen.

File: pkgs/top-level/haskell-packages.nix
Comment (a-m-joseph) [reply pkgs/top-level/haskell-packages.nix:100]: Done.

**Comment by a-m-joseph** [comment]
> Oh, and the commit message(s) should read `haskell.compiler.ghc865Binary: …` for the bindist commits when you squash at the end.

Done

**Comment by a-m-joseph** [comment]
> To clarify, you mean `ghc922`, right? `gh921` was removed a few weeks ago on `master`.

No, it was actually `ghc921`.  My personal fork of nixpkgs (which now supplies the userspace for most of my machines) is based off of `462ebd02ea333a997aff8c787651d9c7a2bcc707` which is now a month old.  I will rebase sometime this week.

**Comment by a-m-joseph** [comment]
> This still means that you have to _keep_ it working, check for regressions, maintain or add platform specific workarounds, …

In nixpkgs, merging architectural conditionals definitely does not imply that level of support.

About half of the work it took me to get nixpkgs working on powerpc64le was already committed before I started.  I'm really glad that stuff was allowed in, even though it didn't amount to working support.  It made the size of my task much more manageable.

> and probably also keeping expectations low until we know we can pull it off

I agree with this, and would say that this is already effectively the unwritten policy for nixpkgs.  Anything I find that works on anything other than `{x86_64,aarch64}-linux` I consider to be an unexpected bonus!

> It's often difficult to tell whether it is still necessary after an update

This is an excellent point, and since about two weeks ago I've begun being very careful to clearly state "this can be removed when..." anywhere I add architecture-specific code that isn't part of a Rosetta Stone.  [Here](https://github.com/NixOS/nixpkgs/pull/161162/files#diff-2828f66a476875b1160e3c241960db085c2a3933211cc187bcd1fb456fb95e7fR120) is one example.  In #166879 I went so far as to [have the workaround automatically insist on itself being removed as soon as it is no longer necessary](https://github.com/NixOS/nixpkgs/pull/166879/files#diff-ab5748dc9567516fefba8344056b51ec1866adeace380f46e58a7af3d619ea22R195).

So yes, I agree, and I would definitely support some kind of official policy that any architecture-specific code other than a Rosetta Stone entry or a bootstrap tarball hash must clearly explain under what conditions it can be removed, and checking those conditions shouldn't require access to hardware.  That would be a good policy to have.

> but due to this hype

:)

> Sadly on classic powerpc big endian it seems, don't remember whether you wanted to tackle that.

Well, next on my list after powerpc64le is big-endian mips64.  I've heard that big-endian is a real headache because so much open source software is developed on little-endian hardware.

If that turns out to be not so bad I can look at big-endian powerpc.  What kinds of powerpc systems are commonly operated in big-endian mode?  I don't remember when powerpc became dual-endian, but I think it was quite a while ago, wasn't it?  Do people run them big-endian due to needing a big-endian kernel for some reason?

**Comment by sternenseemann** [comment]
> If that turns out to be not so bad I can look at big-endian powerpc. What kinds of powerpc systems are commonly operated in big-endian mode?

In my case it was an old iMac I had laying around, I have no clue if any modern and actually used systems use big endian.

**Comment by a-m-joseph** [comment]
@sternenseeman, I have been getting occasional segfaults from the binaries that ghc-8.6.5 produces on powerpc64el-linux.  For example:

```
[749637.456652] ghc:w[17623]: segfault (11) at 4201e58c9000 nip 7ffff2001d74 lr 7fffea03ee68 code 1
[749637.456655] ghc:w[17637]: segfault (11) at 4201e58c9000 nip 7ffff2001d74 lr 7fffea009614 code 1
[749637.456659]  in libHSghc-8.6.5-ghc8.6.5.so[7fffee5e0000+7250000]
[749637.456663]  in libHSghc-8.6.5-ghc8.6.5.so[7fffee5e0000+7250000]
[749637.456665] ghc:w[17623]: code:
[749637.456666] ghc:w[17637]: code:
[749637.456667] 384242c0 3bf6ffc8
[749637.456669] 384242c0
[749637.456670] 7c3fc040 41c0004c
[749637.456671] 3bf6ffc8
[749637.456672] 3fe2fef0 3bff61d8
[749637.456673] 7c3fc040
[749637.456674] fbf6fff0 7dee7b78
[749637.456676] 41c0004c
```

They seem to happen most frequently in the `Setup` binary that ghc tends to compile as part of building a package.  So far I can always get things to continue by simply retrying the build.

I've put this machine through a whole lot of stressful compilation over the last week, and have not encountered any unexpected segfaults from anything other than ghc and/or ghc-compiled code, so I don't think the machine is at fault.  I will have a second Raptor Talos II online in the next few days.

It appears that with the patch above, nix treats ghc-8.6.5 as the "preferred haskell compiler" for powerpc64el.  Is there some way I can tell nixpkgs that it ought to be using something newer?  For example,

```
nix-build . -A haskellPackages.happy
```

... will use the bootstrapping binary to build happy, rather than using a nix-built ghc (which I would imagine is always preferable, no?)

I'm a bit confused about why this is happening, because as far as I can tell `ghc865Binary` is not used for anything except the `bootPkgs` of other ghc derivations.  But it's a bit hard to untangle the control flow of `haskell-packages.nix`...

**Comment by a-m-joseph** [comment]
> I'm a bit confused about why this is happening, because as far as I can tell `ghc865Binary` is not used for anything except the `bootPkgs` of other ghc derivations. But it's a bit hard to untangle the control flow of `haskell-packages.nix`...

Oh, duh, because ghc uses happy to do parsing.  Sorry, I'm an idiot.

I will keep an eye out for any segfaults involving code compiled by a nixpkgs-compiled version of ghc.  I don't think that's happened yet.

It might be that ghc 8.6.5 is not completely stable on powerpc64le, but later versions are.  In that case we might just have to accept a slightly-flaky bootstrap, since ghc doesn't publish any newer binaries.

**Comment by sternenseemann** [comment]
How should we procede here? IUIC you'd not be able to reach this yet due to some compilation failure in `stdenv` bootstrap?

**Comment by a-m-joseph** [comment]
Thanks for remembering to undraftify this!

> you'd not be able to reach this yet due to some compilation failure in `stdenv` bootstrap?

#168199 has merged, so this PR is now usable if you cherry-pick the one-liner in #168983 -- that PR is only draftified because I would feel guilty about asking for it to be merged without first nailing down exactly why it is needed and documenting the reason.  I should have time to take a crack at that on Tuesday.

> How should we procede here?

I'm not sure if you were asking other committers, or me, or all of the above; in any case, I recommend that this be merged.

The only lingering issue here is the [occasional segfaults encountered](https://github.com/NixOS/nixpkgs/pull/168213#issuecomment-1099780754) during the process of using the haskell.org-provided bootstrapping binaries to compile `ghc-8.10.7`.  I have always been able to get past this by simply running `nix-build` a second time.  I have not experienced any problems after `ghc-8.10.7` has been built.  In particular I use `nix-tree` and `nix-diff` very frequently, and both of them work wonderfully.  `nix-tree` in particular performs remarkably well on my powerpc64le machine.

The occasional build needing to be restarted is definitely not ideal, but I think it is way better than nothing at all.  Perhaps a separate bug should be opened noting the flakiness of the ghc865Binary->ghc8107 build step.  In theory I could go digging into this (it has been almost 7 years since I did any work on GHC internals; is it still all System FC?), but I'm not entirely sure what the payback would be if I were to find the root cause.  We'd still need to get the `haskell.org` folks to do a new build for powerpc64le which includes the bugfix... if we can get them to post new tarballs that effort would probably be better spent on posting bootstrap binaries for a newer version of GHC.

Speaking of which, are you in contact with the people who would have the authority to run such a build and post trustworthy bootstrap tarballs?  Could we ask them to use nixpkgs to do this?

> IUIC

Please expand acronym?  Google's best guess is [some kind of religious organization](https://www.google.com/search?hl=en&q=iuic).

**Comment by a-m-joseph** [comment]
Status update: the "correct" fix for the stack-protector issue on powerpc64le is #169378.

However that PR involves a significant change to the stdenv bootstrapping process in order to make the code do what the comments in stdenv/linux/default.nix say it does (which isn't the same thing!).  Merging something like that right now is not appropriate, so it needs to wait until staging reopens for major changes on 2022-May-15.  I've marked #169378 as a draft because I don't want to hassle reviewers during this stressful pre-release period -- especially because the number of people qualified to review that PR is very small.  However the code is ready for review, has gone through a rebuild-the-world on powerpc64le, and should finish a rebuild-the-world on x86_64 tomorrow and aarch64 by the end of this week.

As soon as staging reopens, and people have had time to decompress and rest, I would appreciate your assistance in recruiting reviewers for #169378.  Once that merges, this PR becomes 100% usable.

**Comment by a-m-joseph** [comment]
Latest push was just a rebase, so let's try again:

@ofborg eval

**Comment by sternenseemann** [comment]
I say we merge this after branch-off, as we're targeting post 22.05 staging with the powerpc64le fixes? We can backport it off course if the fixes work out and are as well, wdyt?

**Comment by a-m-joseph** [comment]
> I say we merge this after branch-off, as we're targeting post 22.05 staging with the powerpc64le fixes?

Sounds good to me, I will switch the branch.

**Comment by a-m-joseph** [comment]
@ofborg eval

**Comment by a-m-joseph** [comment]
@ofborg eval

**Comment by a-m-joseph** [comment]
@ofborg eval

**Comment by a-m-joseph** [comment]
>  OfBorg / ghc, ghc.passthru.tests, haskell.compiler.ghc865Binary, haskell.compiler.ghc865Binary.passthru.tests
> on aarch64-darwin Started **54d** 6h 53m 35s ago

I think OfBorg is OfStuck...

**Comment by a-m-joseph** [comment]
@ofborg eval

**Comment by amjoseph-nixpkgs** [comment]
Rebased to resolve merge conflict.

**Comment by amjoseph-nixpkgs** [comment]
> Still waiting on #169378, right?

How embarrassing: my last force-push was from the wrong branch on my end.  I have fixed this.

Is there any way to resolve merge conflicts on github without doing a force-push, and without having to use their text-editor-in-the-web-browser?  Force-pushes are so error-prone...

With `git format-patch`/`git am`, thing being uploaded is a range of commit-hashes rather than just the tip.  I guess github must not have this, because if it did we wouldn't need to set the "base branch" manually via the web UI.

I've started draftifying before each force-push to prevent mistakes from causing mass-pings.

**Comment by amjoseph-nixpkgs** [comment]
Rebased.

**Comment by amjoseph-nixpkgs** [comment]
After using this for the past few months I continue to encounter intermittent segfaults from the `downloads.haskell.org`-supplied GHC binaries for 8.6.5.  The segfaults are extremely nondeterministic.  I've always been able to proceed by simply retrying (never more than 10 attempts).

Once that compiler builds a ghc-8.10.7 from source, everything is completely reliable.

Unfortunately that 8.6.5 binary is the *only* powerpc64le binary I can find on `downloads.haskell.org`.

@sternenseemann, would it be acceptable for me update this expression to `patchelf` a binary from Debian?  If so, are there any constraints I should observe while attempting to make this work, like "must be older/newer than version X"?  My instinct would be to start with the oldest binary I can find which nixpkgs is able to use, in order to maximize the number of GHC versions which can be built from it.

I haven't tried this yet, but wanted to check if there would be any policy obstacles before I sink a bunch of time into this.

**Comment by sternenseemann** [comment]
Using binaries from other distributions for bootstrapping purposes is acceptable and has precedent in nixpkgs (in the case of GNAT).

Usually we try to avoid bootstrapping *chains* to minimize the time we need to spend on bootstrapping a compiler on a stdenv rebuild, since their builds are bottlenecks by nature. However, in this case, I'd say it's up to you – we don't build powerpc64le on Hydra and don't have many support requirements for the platform. Packaging 8.6.* would allow us to build any GHC in nixpkgs after a long chain, whereas packaging 8.10.* would mean that it's much quicker to build the default GHC.

As a side note, you may want to consider making a separate GHC derivation for the debian ones, if there is not much code to be shared with the regular bindists from GHC HQ. I guess in the case of a debian binary, you wouldn't have a configure script etc.

**Comment by amjoseph-nixpkgs** [comment]
> would it be acceptable for me update this expression to `patchelf` a binary from Debian?

So I've been struggling with this and it is not really looking like a net win at the moment.

Debian doesn't package anything between 8.4.4 and 9.0.2, so there won't be any meaningful reduction in the bootstrap-chain length.

Trying to get Debian's binaries to work inside nixpkgs is getting pretty ugly.  I had no idea how many places ghc writes absolute paths into its install... and glibc is not really cooperating (missing `GLIBC_PRIVATE` symbols, etc).

I think leaving the flaky 8.6.5 bootstrap is the best we can do unless GHC HQ is willing to post a blessed build of something more recent.

**Pull request merged**
